Load custom libs from `config/custom_packages`

farmio opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->


## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
Introduce `config/custom_packages` directory to have an easy way of providing (and removing) custom packages (package versions) for development and debug purposes available on any HA installation method.

A repair issue will be raised for each loaded package. 
If HA starts in safe- or recovery-mode, custom packages aren't loaded.

This is an alternative to the previous method loading them from root of config as removed in https://github.com/home-assistant/core/pull/107760 

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [ ] The code change is tested and works locally.
- [ ] Local tests pass. **Your PR cannot be merged unless tests pass**
- [ ] There is no commented out code in this PR.
- [ ] I have followed the [development checklist][dev-checklist]
- [ ] I have followed the [perfect PR recommendations][perfect-pr]
- [ ] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

farmio committed:
Load custom libs from `config/custom_packages`

balloob commented:
What is wrong with installing packages using `pip install -e` ? 

If we were too support such a thing for development, we should limit it to the `--debug` flag I think?

thecode commented:
> What is wrong with installing packages using `pip install -e` ?
> 
> If we were too support such a thing for development, we should limit it to the `--debug` flag I think?

I think there is a huge benefit from being able to directly copy files from your dev machine to a prod or test server without the need to commit them or running pip install. 
I had many situations that I needed to test something on my prod server (otherwise I have to reconfigure the devices to connect to another server which is a pain) and the option for simple copy to a share, restart and test the changes is very effective. 
I understand `pip install -e` will do the same, but it requires an additional step and may be complicated when you want to simply remove the changed files and let your server get back to the original state (since if you use `pip install -e` core will not restore the original dependency as far as I remember if the version did not change).

IMO logging a warning is good enough, I don't see any difference from loading a custom package or loading a custom integration (`custom_components`) both can break core. I understand why https://github.com/home-assistant/core/pull/107760 was needed since it fixed a design flaw that could cause unintentional change that will break something, however this PR suggests an intentional change.

elupus commented:
I kind of like this. This would allow us to remove the deps folder totally. Its mostly legacy.

I think we should just assume our environment is writable normally.  Be it venv or docker environments. Block other startups early in startup.

This would still allow custom deps for testing.

frenck commented:
> Being able to ask a user to add an updated lib to test some unreleased change.

This is also something I'm against. They will run into surprises when they set and forget about it.

elupus commented:
> > Being able to ask a user to add an updated lib to test some unreleased change.
> 
> This is also something I'm against. They will run into surprises when they set and forget about it.

While true. Its not always possible to have access to all users systems. Some users might have to be guinipigs

thecode commented:
> > Being able to ask a user to add an updated lib to test some unreleased change.
> 
> This is also something I'm against. They will run into surprises when they set and forget about it.

While I agree many users will forget about it, I don't see how it is different that when developers ask a user to install a change as a custom integration, but I was aiming for support for developers anyhow and not for users testing so it is possible to think of how to limit it (although not ideal, Paulus suggestion only with debug flag is also acceptable).

I think similar to custom integrations if something is loaded from `custom_packages` folder it should log a warning include it in diagnostics and even mark the installation as unsupported, this should reduce surprises

farmio commented:
Hi 👋! 
Sorry for the late reply, I'm on holidays ⛷️😃

My original intention and use-cases are following:
- allow developers to quickly add and remove libs with their local changes. Also add and remove multiple of them (via symlinks etc.) is a matter of seconds with a custom_packages directory. Even if the libs are not ready or not suitable for an editable pip install (I personally blew my venv setup trying to do that last week. I have lib using a dynamic version key in pyprojcet.toml - this is set by a GH-Action on release. I know this can be fixed, but I can't guarantee that others won't run into the same problem...)

- allow users to test custom made lib versions for *very special problems*. This shouldn't be the major use of this feature, but sometimes there are users with who can just reproduce bugs with their specific hardware setups, that can't be reproduced by a dev. In this case a simple copy&paste of files - and deleting them later - is imho just the easiest way to set them up with a custom lib version. I see the issue of "forgetting to remove" the lib, but that's where the log warning and not loading in safe-mode enters the game (its the same as done for custom components). But maybe we can even create a repair issue or something like that.

balloob commented:
I think Frencks concern is the big one for me too. Users will put it up, it works and then upgrade Home Assistant, and it can take Home Assistant down because something was overwritten, they forgot about it and imports now fail (something we can't catch). 

Is there a way we can make this temporary and only meant fo testing? We want to avoid people putting this in their instance as a bug-fix for that version, instead of waiting for a patch-release. This will go wrong and people will blame us.

thecode commented:
> I think Frencks concern is the big one for me too. Users will put it up, it works and then upgrade Home Assistant, and it can take Home Assistant down because something was overwritten, they forgot about it and imports now fail (something we can't catch).
> 
> Is there a way we can make this temporary and only meant fo testing? We want to avoid people putting this in their instance as a bug-fix for that version, instead of waiting for a patch-release. This will go wrong and people will blame us.

I think the concern of breaking up core, mainly after update is a valid claim. TBH I just noticed my prod is running a custom integration since I forgot to remove it after testing. Since @farmio also indented this would be used by users for testing the debug flag is not an option.

If our only concern is core update maybe we can include a manifest that specifically states which core version the custom lib should be used on, since the PR description limit it for specific user testing and not wide testing it might be possible to include either a manifest for the dependency or include it on the integration which is used to be debugged, so that once core is updated it will no longer be loaded (not sure how easy to implement).

I think the first step is to agree that if this is limited somehow the option add custom libs is acceptable than we can discuss how to limit it (I also had in mind to limit it to a single run, so it is removed after restart, but might be too dangerous to do it)

farmio committed:
Raise repair issue

adding core version to issue_id so it re-raises after update if ignored previously

farmio commented:
I've added the creation of a repair issue, reissued when Core was updated to avoid ignores, for each custom_package.

If adding this `custom_packages` feature is out of question, just let me know - it was not clear from previous discussion.

farmio committed:
Remove custom_packages from path in safe_mode

farmio committed:
Tests

emontnemery commented:
We discussed this in the core team and decided we won't accept this PR at this point.
